src/ILVerify/src/Verifier.cs

```diff
@@ -0,0 +1,224 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Reflection;
+using System.Reflection.PortableExecutable;
+using System.Resources;
+using System.Text;
+using Internal.IL;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace ILVerify
+{
+    public delegate bool ShouldVerifyMethod(string name);
+
+    public interface IResolver
+    {
+        PEReader Resolve(AssemblyName name);
+    }
+
+    public class Verifier
+    {
+        private Lazy<ResourceManager> _stringResourceManager =
+            new Lazy<ResourceManager>(() => new ResourceManager("ILVerify.Resources.Strings", Assembly.GetExecutingAssembly()));
+
+        private SimpleTypeSystemContext _typeSystemContext;
+
+        public ShouldVerifyMethod ShouldVerifyMethod { private get; set; }
+
+        public Verifier(IResolver resolver)
+        {
+            _typeSystemContext = new SimpleTypeSystemContext(resolver);
+        }
+
+        internal Verifier(SimpleTypeSystemContext context)
+        {
+            _typeSystemContext = context;
+        }
+
+        public void SetSystemModuleName(AssemblyName name)
+        {
+            _typeSystemContext.SetSystemModule(_typeSystemContext.GetModule(name));
+        }
+
+        public VerificationResult Verify(AssemblyName moduleToVerify)
```

Comment:
> I agree. As discussed in the corresponding issue, it would make sense to me, to have:
> Verify(PEReader) for verifying entire modules
> Verify(PEReader, MethodDefinitionHandle) for verifying a specific method
> Verify(PEReader, TypeDefinitionHandle) for verifying a specific type

If the API accepts `PEReader`, then we have to give up caching in `SimpleTypeSystemContext`, since `PEReader`s can't be keys in a dictionary. New `EcmaModule`s will be recreated multiple times.
I don't know if that is just a performance concern or if that might cause identity problems as well...

Reply:
> since PEReaders can't be keys in a dictionary

Why not? We can use their object identity.

Reply:
The `PEReader`s returned by `IResolver.Resolve` are currently not cached, so we're getting new ones. I'll add caching then object identity should work.